ci-operator/step-registry/cucushift/e2e/cucushift-e2e-commands.sh

```diff
@@ -40,12 +40,44 @@ function echo_e2e_tags() {
 }
 function filter_test_by_version() {
     local xversion yversion
-    IFS='.' read xversion yversion _ < <(oc version -o yaml | yq '.openshiftVersion')
+    OPENSHIFT_VERSION="$(oc version -o yaml | yq '.openshiftVersion')"
+    IFS='.' read xversion yversion _ <<< "$OPENSHIFT_VERSION"
     if [[ -n $xversion ]] && [[ $xversion -eq 4 ]] && [[ -n $yversion ]] && [[ $yversion =~ [12][0-9] ]] ; then
         export E2E_RUN_TAGS="${E2E_RUN_TAGS} and @${xversion}.${yversion}"
     fi
     echo_e2e_tags
 }
+function filter_test_by_arch() {
+    local version node_archs
+    version="$OPENSHIFT_VERSION"
+    if [[ -n "$version" ]] ; then
+        if [[ "$version" =~ multi ]] ; then
+            mapfile -t node_archs < <(oc get nodes -o yaml | yq '.items[].status.nodeInfo.architecture' | sort -u)
+            case "${#node_archs[@]}" in
+                1)
+                    export E2E_RUN_TAGS="${E2E_RUN_TAGS} and @${node_archs[0]}"
+                    ;;
+                2)
+                    export E2E_RUN_TAGS="${E2E_RUN_TAGS} and @heterogeneous"
```

Comment:
hi Liang, thanks for this work.
As we approach more architectures for testing, I wonder if we can also add the node_archs in the heterogeneous case. Which is: if the nodes architectures are `arm64` and `amd64`, let's add
`@heterogeneous @arm64 @amd64`.

This way, we execute tests that (1) can run on heterogeneous clusters, and (2) on both arm64 or amd64 nodes.

For example, a test case missing arm64 but having ppc64le and heterogeneous will not be selected.

Finally, I'd not assume 2 is the maximum number of architectures... In the future, we might have confs like arm64 CP + amd64 & ppc64le workers.

Reply:
The current design is based on what we already have. I think we can update the logic later if we have such CP+worker combinations in the future.

Reply:
I'd insist for the additional tags in the heterogenous case at least.

For the more than 2 architectures, as it should be trivial I'd it too (the prow steps are almost ready to onboard that change now, and avoiding another pr here is "nice to have")

Reply:
You mean for the heterogenous cases, you would like it to be `export E2E_RUN_TAGS="${E2E_RUN_TAGS} and @heterogeneous and @amd64 and @arm64`, right ?

Reply:
For the more than 2 architectures, I'd prefer a follow up PR for that. If we add it now, the logic for that part can not be tested in real tests as we don't have that now.

Reply:
> You mean for the heterogenous cases, you would like it to be export E2E_RUN_TAGS="${E2E_RUN_TAGS} and @heterogeneous and @amd64 and @arm64, right ?

Yes, this was already on the queue. We want to ensure not only that a test can run on heterogeneous clusters, but on heterogeneous clusters with specific configurations. Some cases might be ok to run on heterogeneous clusters with P/X, but not A/X clusters. A similar change is required for some functions in openshift-tests-private and we'd work on that too eventually.

> For the more than 2 architectures, I'd prefer a follow up PR for that. If we add it now, the logic for that part can not be tested in real tests as we don't have that now.

As you prefer, from my perspective, it would mean changing the `2)` to `*)` and removing the current default case that fails with more than 2. As of now, once we onboard a three architectures cluster test-config, cucushift will fail on the default branch of the switch statement you're introducing.

Reply:
> I'd insist for the additional tags in the heterogenous case at least.

Done

Reply:
some script like this should resolve the first comment:
```shell
#!/bin/bash

join_by_prefix() {
  local delimiter="${1}"
  shift
  local array=("$@")
  local result=""
  local i
  for ((i=0; i<${#array[@]}; i++)); do
    result+="${delimiter}${array[i]}"
  done
  echo -n "$result"
}
my_array=("amd64" "ppc64le" "arm64")
echo "${my_array[@]}"
join_by_prefix " @" "${my_array[@]}"
echo

my_array=("amd64")
echo "${my_array[@]}"
join_by_prefix " @" "${my_array[@]}"
echo

my_array=("amd64" "ppc64le")
echo "${my_array[@]}"
join_by_prefix " @" "${my_array[@]}"
echo

my_array=()
echo "${my_array[@]}"
join_by_prefix " @" "${my_array[@]}"
echo
```